### PR TITLE
delpay: introduced a new rpc method to delete a payment by hash

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Hi all.

This Draft implement a new RPC command to delete the payment with the status `complete` `failed` marked like good first issue https://github.com/ElementsProject/lightning/issues/1995

However, this PR depends from the PR where I proposed to update the result of listpays command with `payment_hash` inside the result https://github.com/ElementsProject/lightning/pull/3888. In addition, with the @cdecker suggestion to add the `payment_hash` in all cases of types of payments, the command `delpay` can use only the `payment_hash` to identify the payment to remove and not `bolt11`.

This PR is a draft yet because I have some point to clarify with the team, for instance, the type of error that the command should be returned in cases of error. In other words, I see that the `LIGHTNINGD` error is marked with a `FIXME` because each command should use the correct error code. I'm not able to find the payment error codes inside the code. maybe we could define the payment error code?

- [x] Docs update with the change proposed in this Draft
- [X] Fix error code inside the command implementation

Changelog-Added: JSON-RPC: delpay a new method to delete the payment completed or failed.

### CODE REVIEW

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [inline doc/lightning-delpay.7.md:7]: You have a `partid` as well that is not documented.

Defaults should be described in "DESCRIPTION" section, not in the synopsis.

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [inline doc/lightning-delpay.7.md:14]: DESCRIPTION section needs to describe every parameter, and describe defaults for that parameter as well if it is optional and not specified.

**Comment by ZmnSCPxj** [comment]
In addition, it seems to me better to just delete *all* `partid`s, not delete them one-by-one.  Loss of one `partid` of a multi-part payment is already a loss of information anyway, so just delete them all in one database transaction.

**Comment by vincenzopalazzo** [comment]
Hi @ZmnSCPxj, thanks to review this PR.

I will make the changes inside the docs and inside the JSON RPC errors, sorry about that.

In addition, about your suggestion.

>In addition, it seems to me better to just delete all partid, not delete them one-by-one. Loss of one partid of a multi-part payment is already a loss of information anyway, so just delete them all in one database transaction.

For my code comprehension, to make this work, so, to delete the complete payment and not each payment by `partid`, I need a new function inside the wallet.c to run the query "DELETE FROM payments WHERE payment_hash=?"

I'm losing something?

Maybe I can use the `function wallet_payment_list` and run for each element inside the list `wallet_payment_delete` but I think that the first solution is cleaner but I don't know if it is safe to run on the lightning node the delete query by `hash`?

**Comment by ZmnSCPxj** [comment]
`DELETE FROM TABLE payments WHERE payment_hash=?` should be fine.  You might want to do some check to check if this results in any changes (I think we have a function to check the number of changes a query makes) and if it is 0 that is probably fine.

But you want to check the state of the *entire* payment first. The logic for a purported "payment state":

* If any part has succeeded, then the entire payment succeeded.
* If any part is pending, then the entire payment is pending.
* Otherwise, all parts are failed, so the entire payment is failed.

It is safe to delete a payment if the entire payment is failed or succeeded, and we should probably disallow deleting pending payments (which might still be operated by `pay` over in the `pay` plugin).

Note that if you do not pass control to a function with a callback, then the entire processing is done inside a database transaction, so you are assured that between the check and the deletion, the database state is consistent.  But if you pass control to a function with a callback, the database transaction will end at that point, and by the time the callback is called, you are in a new transaction.  So do not call into any callback-using functions between checking the state of the payment and deleting the whole payment.

**Comment by vincenzopalazzo** [comment]
Thanks for the suggestion @ZmnSCPxj.

>You might want to do some check to check if this results in any changes (I think we have a function to check the number of changes a query makes) and if it is 0 that is probably a "payment not found" error.

I make the sanity test of input before to call the database function, but I can improve the check maybe I can make the following test:

1. payment not found: I used the `wallet_payment_by_hash` to find the payment, I think that is possible use this function to get the payment with hash and *partid=0*, if exist this part, the payment exists, right?

2. payment status not valid: from your previous message I understand the payment status is decided from the status of all part, so, I should check the status of all payment find with `wallet_payment_list` and check the status for each payment and make the correct decision.

of course, there are all checks on the input but, they are not relevant into this description.

>It is safe to delete a payment if the entire payment is failed or succeeded, and we should probably disallow deleting pending payments (which might still be operated by pay over in the pay plugin).

Yep, I agree, with my actual logic inside the `delpay` command you can remove only the payment with the status `complete` or `failed`, but of course, with your suggestion the logic change a little.

**Comment by ZmnSCPxj** [comment]
`partid` 0 is for non-split payments.  If we split payments at all, there might not be a `partid` 0.

**Comment by vincenzopalazzo** [comment]
Rebased https://github.com/ElementsProject/lightning/commit/01a82d38f769a410e4f2d127c9f09e8f8b296554 and updated https://github.com/vincenzopalazzo/lightning/commit/3a5f0be11e8f9efe6bc7796a0fd3289e83aebc52

File: doc/lightning-delpay.7.md
Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md]: Updaed

File: doc/lightning-delpay.7.md
Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md:17]: Updated but the doc needs a second review in the next hours (Caused from my Dyslexia problem :smile: )

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [inline doc/lightning-delpay.7.md:16]: You should list here the possible values.

```
*status* is the expected status of the payment.
It can be *complete* or *pending*.
Only delete if the payment status matches.
If not specified, defaults to *complete*.
```

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [inline doc/lightning-delpay.7.md:43]:
```
*payment\_hash*
```

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [inline doc/lightning-delpay.7.md:7]: Escape the underscore (as well all instances)

```
*payment\_hash*
```

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:1542]: Check is unnecessary, `p_req` will check already.

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:1538]: Can use instead:

```c
p_opt_def("status", param_string, &status_str, "complete")
```

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:1547]: See above for `p_opt_def`.

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:1551]: "status insert"?  Better "Unrecognized status: %s".

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:45]: The usual convention for parsers that can fail is something like this:

```c
bool string_to_payment_status(const char *status_str, enum wallet_payment_status *)
```

Then later in use:

```
if (!string_to_payment_status(status_str, &status))
    return command_fail(cmd, JSONRPC2_INVALID_PARAMS, "Unrecognized status '%s'", status_str);
```

File: wallet/wallet.c
Comment (ZmnSCPxj) [inline wallet/wallet.c:2690]: Looks correct but @cdecker should check as well, I am no longer familiar with how we interact with the db.

File: lightningd/pay.c
Comment (ZmnSCPxj) [inline lightningd/pay.c:1554]: This is an error in the argument, not in the status of the actual payment, so should be `JSONRPC2_INVALID_PARAMS`.

**Review by ZmnSCPxj — DISMISSED** [review]

File: wallet/wallet.c
Comment (cdecker) [reply wallet/wallet.c:2690]: This should work just fine. The SQL subsystem just takes a valid query and rewrites it for various dialects. That works well for this type of simple queries, but may be a bit more complex for things like upserts or last inserted IDs, which we use in some places.

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c:1538]: @ZmnSCPxj, mh! maybe I'm losing something here, param_string accept only a `const char *str` and the function `o_opt_def` can not assign a new value at `status_str` variable.

With this comment, you mean that I need to declare some function that accepts a not const `char*`?

I think that I'm losing something, sorry about that

Comment (ZmnSCPxj) [reply lightningd/pay.c:1538]: String constants *are* `const char *`.  You can do something like `const char *foo = "foo";`.  You can even do something like `const char *foo = "foo"; foo = "bar"`.

The *correct* way to write `const` is *after* the type.  So actually `const char *` should be written as `char const *`.  Then you read types from right to left.  So `char const *` is read as "pointer (`*`) to constant (`const`) character (`char`)".  Note that the pointer itself is not `const`, so you can modify the pointer, which is why `char const* foo; foo = "bar"; foo = "quux"` is valid.  What you cannot do is `char const* foo = "foo"; *foo = '0';`, because what is pointed *to* is constant.

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c]: Changed in https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c]: Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c]: Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c]: Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: doc/lightning-delpay.7.md
Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md]: Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: doc/lightning-delpay.7.md
Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md]: Changed https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: doc/lightning-delpay.7.md
Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md]: Thanks for this suggestion. modify made into https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c:1541]: Thanks for this wonderful informations. I made a couple of change inside https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934 but I want to have more detail about these changes. I like your suggestion to use p_opt_def.

File: lightningd/pay.c
Comment (vincenzopalazzo) [inline lightningd/pay.c:45]: @ZmnSCPxj I understood your suggestions inside this function? or I'm introducing some mistake?

Comment (ZmnSCPxj) [reply lightningd/pay.c:45]: looks ok, but no time for in-depth review now, please wait....

File: tests/test_pay.py
Comment (niftynei) [inline tests/test_pay.py:3317]: this test could easily be rolled into the previous one

File: lightningd/pay.c
Comment (niftynei) [inline lightningd/pay.c:1562]: line this up with `cmd` on the previous line?

File: lightningd/pay.c
Comment (niftynei) [inline lightningd/pay.c:1552]: needs to be aligned with `cmd`

File: lightningd/pay.c
Comment (niftynei) [inline lightningd/pay.c:1568]: align with `cmd`?

File: lightningd/pay.c
Comment (niftynei) [inline lightningd/pay.c:67]: it's best practice not to use `default` for a switch statement. instead, name all of the cases.

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:13]:
```suggestion
status. Deleting a `pending` payment is an error.
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:15]:
```suggestion
- *payment\_hash*: The unique identifier of a payment. Find in **listpays** or **listsendpays**;
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:16]:
```suggestion
- *status*: Expected status of the payment. Valid values are *complete* or *failed*. 
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:17]:
```suggestion
Only deletes if the payment status matches. If not specified, defaults to *complete*.
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:35]:
```suggestion
On success, the command will return a payment object, such as the **listsendpays**. If the payment is an MPP (multi-part payment) a list of 
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:36]:
```suggestion
payments will be returned -- one payment object for each partid.
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:38]:
```suggestion
On failure, an error is returned. If the lightning process fails before responding, the
```

we don't delete the payment on failure?

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:43]:
```suggestion
- -32602: Parameter missing or malformed;
```

File: doc/lightning-delpay.7.md
Comment (niftynei) [inline doc/lightning-delpay.7.md:44]:
```suggestion
- 211: Payment status mismatch. Check the correct status via **paystatus**;
```

**Comment by niftynei** [comment]
Mostly grammatical/spelling fixes

**Comment by vincenzopalazzo** [comment]
Rebase and update with suggestions.

Thanks @niftynei about the fixes and sorry for the spelling errors. I'm still searching somethings where I'm good to do but I'm sure that I'm very bad to write sentences :smile:

File: doc/lightning-delpay.7.md
Comment (cdecker) [inline doc/lightning-delpay.7.md:13]:
```suggestion
The **delpay** RPC command deletes a payment with the given `payment_hash` if its status is either `complete` or `failed`. Deleting a `pending` payment is an error.
```

File: doc/lightning-delpay.7.md
Comment (cdecker) [inline doc/lightning-delpay.7.md:17]: Shouldn't an unspecified value for `status` mean we don't care, and either `complete` or `failed` will be removed?

File: doc/lightning-delpay.7.md
Comment (cdecker) [inline doc/lightning-delpay.7.md:35]:
```suggestion
If successful the command returns a payment object, in the same format as **listsendpays**.
If the payment is a multi-part payment (MPP) the command return a list of 
```

File: lightningd/pay.c
Comment (cdecker) [inline lightningd/pay.c:1557]: Probably it's better to check if we have one of the valid statuses instead of not having the third. If we ever add a new status this is would not cover that fourth state. If you use a `switch` statement then the compiler will make sure we cover all cases exhaustively.

File: wallet/wallet.c
Comment (cdecker) [inline wallet/wallet.c:2672]: Spurious whitespace change.

File: wallet/wallet.h
Comment (cdecker) [inline wallet/wallet.h:1036]: Nit: indentation of the followup line should match the `struct` keyword above (might be a GH display issue though).

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3266]:
```suggestion
    This test includes all possible combinations of input error inside the
```

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3279]: Well, really this matches the "no payment yet" case, since `l2` doesn't know anything about that payment. So you could just use `payment_hash = "AA"*32` here.

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3283]: Payment was no attempted at all, so it'd fail for that reason already.

Comment (cdecker) [reply tests/test_pay.py:3283]: Basically this is testing our sanity check order, and not the underlying `pending` vs. `non-pending` cases.

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3289]: You could have used `l1, l2 = node_factory.line_graph(2)` above since so far the channel has not been used. That'd save a couple of test-lines

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3294]: These would have also been subsumed by `line_graph`

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3302]: This is redundant with the test on line 3283

File: tests/test_pay.py
Comment (cdecker) [inline tests/test_pay.py:3323]: Wumbo is not needed here: `MPP_TARGET_SIZE` is denominated in msats, whereas the `fundamount` is denominated in sats. So this creates a channel that is 4 orders of magnitude larger than the amount being transferred:

 - Channel capacity: 10**8 * 1000 = 10**11
 - Payment amount: 5 * 10**7

Doesn't hurt though :wink:

File: doc/lightning-delpay.7.md
Comment (ZmnSCPxj) [reply doc/lightning-delpay.7.md:17]: Since `delpay` is supposed to mirror `delinvoice`, which has the `*status*` as a required arg, maybe it should be a required arg here as well?

Comment (vincenzopalazzo) [reply doc/lightning-delpay.7.md:17]: Yep I agree with this point! thanks for clarifying!

File: lightningd/pay.c
Comment (vincenzopalazzo) [reply lightningd/pay.c:1557]: Thanks for this suggestion, I made the change from if to switch!

File: wallet/wallet.c
Comment (vincenzopalazzo) [reply wallet/wallet.c:2672]: Ops! thanks

File: tests/test_pay.py
Comment (vincenzopalazzo) [reply tests/test_pay.py:3283]: This comment brings me to reflect on the sens of this test!

I moved the pay command to the top of this sanity check, so, with this new situation the unit test has more sense to me because it does the sanity check and the command behavior checks

File: tests/test_pay.py
Comment (vincenzopalazzo) [reply tests/test_pay.py]: In these cases, I could be used the following line of code to have the channels ready to work `node_factory.line_graph(2, fundamount=10**5, wait_for_announce=True)` , right?

File: tests/test_pay.py
Comment (vincenzopalazzo) [reply tests/test_pay.py]: >Doesn't hurt though

Yes, it is, but it is also my logic error, I discovered only now that `fundamount` is denominated in sats. I think that `10**5` is enough

**Comment by vincenzopalazzo** [comment]
Rebase and update with review https://github.com/ElementsProject/lightning/commit/3314a059779c4769771a4d086b772b9c40827c14

Waiting to Travis test result

**Comment by cdecker** [comment]
Rebased on top of `master`

**Review by cdecker — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/3899/commits/56a5dc93d43bbadaa64b5be9dd0c07189f5dab27

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/pull/3899/commits/db5323522e23bb5f918ec0d96790316212758621

**Pull request merged**
